### PR TITLE
vim - add command for grep

### PR DESCRIPTION
**Pull request by wuphilipp**
*  add a `:Grep` command for searching with fzf

 * add a bind to grep for a visual selection

 * add a binding to grep for the current word

### CODE REVIEW

**Review by brentyi — APPROVED** [review]
looks ok to me!
@wuphilipp can you take a look at the changes I made before merging? notably replaced `git grep` with `ag` and tweaked bindings slightly. downside is that hidden files are currently not searched, but it seems to work better for big binaries etc for me. also a bit more versatile (outside of repos, in mercurial, etc)

File: common/.vimrc [hunk 1]
```diff
@@ -183,15 +183,48 @@ else
             \         . ' ' . s:preview_script . ''''
             \ }, <bang>0)
 
-        " Bindings
+        " Using grep for visual mode selection
```
Comment (wuphilipp) [inline common/.vimrc:192]:
```suggestion
        " Using grep for visual mode selection
        command! -bang -nargs=* Ag
            \ call fzf#vim#ag(<q-args>, fzf#vim#with_preview({
            \     'dir': b:repo_file_search_root
            \ }), <bang>0)

```

File: common/.vimrc [hunk 2]
```diff
@@ -183,15 +183,48 @@ else
...
         nnoremap <Leader>gl :call fzf#vim#lines(expand('<cword>'))<CR>
-        nnoremap <Leader>gf :call fzf#vim#files(b:repo_file_search_root, {
-            \ 'options': '--query ' . expand('<cfile>')})<CR>
+
+        " Bindings: search lines in files with ag
+        nnoremap <Leader>a :call fzf#vim#ag('', fzf#vim#with_preview({
+            \ 'dir': b:repo_file_search_root}))<CR>
+        vnoremap <Leader>a :<c-u>call <SID>GrepVisual(visualmode())<cr>
+        nnoremap <Leader>ga :call fzf#vim#ag('', fzf#vim#with_preview({
+            \ 'options': '--query ' . shellescape(expand('<cword>')), 'dir': b:repo_file_search_root}))<CR>
```
Comment (wuphilipp) [inline common/.vimrc]:
```suggestion
        nnoremap <Leader>a :Ag<CR>
        vnoremap <Leader>a :<c-u>call <SID>GrepVisual(visualmode())<cr>
        nnoremap <Leader>ga :execute 'Ag ' . expand('<cword>')<CR>
```

File: common/.vimrc [hunk 3]
```diff
@@ -183,15 +183,48 @@ else
...
+                normal! `<v`>y
+            elseif a:type ==# 'char'
+                normal! `[v`]y
+            else
+                return
+            endif
+
+            call fzf#vim#ag('', fzf#vim#with_preview({
+                \ 'options': '--query ' . shellescape(@@),
+                \ 'dir': b:repo_file_search_root}))
```
Comment (wuphilipp) [inline common/.vimrc]:
```suggestion
execute "Ag " @@
```

**Review by wuphilipp — COMMENTED** [review]
Could we keep an actual Ag: command to run for convenience? I made some suggestions.

This also behaves slightly differently. For example, <Leader ga> will run Ag on the current word. Then you can fzf the result. (as opposed to running ag on everything then fzf the current word).

**Comment by brentyi** [comment]
@wuphilipp what's the benefit of running `:Ag (initial search)`, as opposed to pre-populating the query?

**Comment by brentyi** [comment]
@wuphilipp Ah that makes sense -- just accepted your suggestions + added support for hidden files

**Pull request merged**